File: library/src/main/java/com/afollestad/aesthetic/views/AestheticActionMenuItemView.java
package com.afollestad.aesthetic.views;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.support.annotation.RestrictTo;
import android.support.v7.view.menu.ActionMenuItemView;
import android.util.AttributeSet;

import com.afollestad.aesthetic.Aesthetic;

import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;

import static android.support.annotation.RestrictTo.Scope.LIBRARY_GROUP;
import static com.afollestad.aesthetic.Rx.distinctToMainThread;
import static com.afollestad.aesthetic.Rx.onErrorLogAndRethrow;
import static com.afollestad.aesthetic.TintHelper.createTintedDrawable;
import static com.afollestad.aesthetic.Util.isColorLight;

@SuppressWarnings("RestrictedApi")
@RestrictTo(LIBRARY_GROUP)
public class AestheticActionMenuItemView extends ActionMenuItemView {

  private Drawable icon;
  private Subscription subscription;

  public AestheticActionMenuItemView(Context context) {
    super(context);
  }

  public AestheticActionMenuItemView(Context context, AttributeSet attrs) {
    super(context, attrs);
  }

  public AestheticActionMenuItemView(Context context, AttributeSet attrs, int defStyle) {
    super(context, attrs, defStyle);
  }

  @Override
  public void setIcon(final Drawable icon) {
    // We need to retrieve the color again here.
    // For some reason, without this, a transparent color is used and the icon disappears
    // when the overflow menu opens.
    Aesthetic.get()
        .primaryColor()
        .observeOn(AndroidSchedulers.mainThread())
        .take(1)
        .subscribe(
            color -> {
              int iconColor = isColorLight(color) ? Color.BLACK : Color.WHITE;
              this.icon = createTintedDrawable(icon, iconColor);
              super.setIcon(this.icon);
            },
            onErrorLogAndRethrow());
  }

  @Override
  protected void onAttachedToWindow() {
    super.onAttachedToWindow();
    subscription =
        Aesthetic.get()
            .primaryColor()
            .compose(distinctToMainThread())
            .subscribe(
                color -> {
                  if (icon != null) {
                    setIcon(icon);
                  }
                },
                onErrorLogAndRethrow());
  }

  @Override
  protected void onDetachedFromWindow() {
    subscription.unsubscribe();
    super.onDetachedFromWindow();
  }
}

File: library/src/main/java/com/afollestad/aesthetic/views/AestheticActionMenuItemView.kt
/**
 * Designed and developed by Aidan Follestad (@afollestad)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.afollestad.aesthetic.views

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import androidx.appcompat.view.menu.ActionMenuItemView
import com.afollestad.aesthetic.Aesthetic.Companion.get
import com.afollestad.aesthetic.utils.adjustAlpha
import com.afollestad.aesthetic.utils.one
import com.afollestad.aesthetic.utils.subscribeTo
import com.afollestad.aesthetic.utils.tint
import com.afollestad.aesthetic.utils.toMainThread
import com.afollestad.aesthetic.utils.unsubscribeOnDetach

/** @author Aidan Follestad (afollestad) */
@SuppressLint("RestrictedApi")
internal class AestheticActionMenuItemView(
  context: Context,
  attrs: AttributeSet? = null
) : ActionMenuItemView(context, attrs) {

  companion object {
    const val UNFOCUSED_ALPHA = 0.5f
  }

  private var icon: Drawable? = null

  private fun invalidateColors(color: Int) {
    val sl = ColorStateList(
        arrayOf(
            intArrayOf(-android.R.attr.state_selected),
            intArrayOf(android.R.attr.state_selected)
        ),
        intArrayOf(
            color.adjustAlpha(UNFOCUSED_ALPHA),
            color
        )
    )
    if (icon != null) {
      setIcon(icon!!, sl)
    }
    setTextColor(color)
  }

  override fun setIcon(icon: Drawable) {
    super.setIcon(icon)
    // We need to retrieve the color again here.
    // For some reason, without this, a transparent color is used and the icon disappears
    // when the overflow menu opens.
    get().toolbarIconColor()
        .one()
        .toMainThread()
        .subscribeTo(::invalidateColors)
        .unsubscribeOnDetach(this)
  }

  @Suppress("MemberVisibilityCanBePrivate")
  fun setIcon(
    icon: Drawable,
    colors: ColorStateList
  ) {
    this.icon = icon
    super.setIcon(icon.tint(colors))
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    get().toolbarIconColor()
        .one()
        .toMainThread()
        .subscribeTo(::invalidateColors)
        .unsubscribeOnDetach(this)
  }
}
